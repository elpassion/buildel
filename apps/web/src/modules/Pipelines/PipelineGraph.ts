import { produce } from 'immer';
import { IBlockConfig, IIO } from './pipelines.hooks';

export interface IPipelineConfig {
  blocks: IBlock[];
}

type IBlock = IBlockConfig;

export function getBlocks(pipeline: IPipelineConfig): IBlock[] {
  return pipeline.blocks;
}

export function connectIO(
  pipeline: IPipelineConfig,
  source: { block: IBlock; output: IIO },
  target: { block: IBlock; input: IIO },
) {
  return produce(pipeline, (draft) => {
    if (source.output.type !== target.input.type) return;
    const inputBlockIndex = draft.blocks.findIndex(
      (block) => block.name === target.block.name,
    );
    if (inputBlockIndex === -1) return;
    draft.blocks[
      inputBlockIndex
    ].opts.input = `${source.block.name}:${source.output.name}`;
  });
}

export function disconnectIO(
  pipeline: IPipelineConfig,
  _source: { block: IBlock; output: IIO },
  destination: { block: IBlock; input: IIO },
) {
  return produce(pipeline, (draft) => {
    const inputBlockIndex = draft.blocks.findIndex(
      (block) => block.name === destination.block.name,
    );
    if (inputBlockIndex === -1) return;
    delete draft.blocks[inputBlockIndex].opts.input;
  });
}

export function getNodes(pipeline: IPipelineConfig) {
  return pipeline.blocks.map((block) => ({
    id: block.name,
    type: block.block_type.type,
    position: { x: 0, y: 0 },
    data: block,
  }));
}

export function getEdges(pipeline: IPipelineConfig) {
  return pipeline.blocks
    .filter((block) => block.opts.input)
    .map((block) => ({
      id: `${block.opts.input}-${block.name}:input`,
      source: block.opts.input.split(':')[0],
      sourceHandle: block.opts.input.split(':')[1],
      target: block.name,
      targetHandle: 'input',
    }));
}

export function removeBlock(pipeline: IPipelineConfig, block: IBlock) {
  return pipeline;
}

export function addBlock(pipeline: IPipelineConfig, block: IBlock) {
  return pipeline;
}

export function updateBlock(pipeline: IPipelineConfig, block: IBlock) {
  return pipeline;
}
