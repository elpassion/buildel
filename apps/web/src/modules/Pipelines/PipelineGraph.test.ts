import { describe, expect, test } from 'vitest';
import { IBlockConfig } from '.';
import * as PipelineGraph from './PipelineGraph';

test(PipelineGraph.getBlocks, () => {
  const pipeline: PipelineGraph.IPipelineConfig = {
    blocks: [textInputBlockConfig],
  };
  expect(PipelineGraph.getBlocks(pipeline)).toEqual([textInputBlockConfig]);
});

describe(PipelineGraph.connectIO, () => {
  test('connecting nodes of the same type works', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [textInputBlockConfig, textOutputBlockConfig],
    };
    expect(
      PipelineGraph.connectIO(
        pipeline,
        {
          block: textInputBlockConfig,
          output: textInputBlockConfig.block_type.outputs.at(0)!,
        },
        {
          block: textOutputBlockConfig,
          input: textOutputBlockConfig.block_type.inputs.at(0)!,
        },
      ),
    ).toEqual({
      blocks: [
        textInputBlockConfig,
        {
          ...textOutputBlockConfig,
          opts: { input: `${textInputBlockConfig.name}:output` },
        },
      ],
    });
  });

  test('connecting nodes of different types does not work', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [textInputBlockConfig, textOutputBlockConfig],
    };
    expect(
      PipelineGraph.connectIO(
        pipeline,
        {
          block: audioInputBlockConfig,
          output: audioInputBlockConfig.block_type.outputs.at(0)!,
        },
        {
          block: textOutputBlockConfig,
          input: textOutputBlockConfig.block_type.inputs.at(0)!,
        },
      ),
    ).toEqual(pipeline);
  });
});

describe(PipelineGraph.disconnectIO, () => {
  test('disconnecting nodes works', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [
        textInputBlockConfig,
        {
          ...textOutputBlockConfig,
          opts: { input: `${textInputBlockConfig.name}:output` },
        },
      ],
    };
    expect(
      PipelineGraph.disconnectIO(
        pipeline,
        {
          block: textInputBlockConfig,
          output: textInputBlockConfig.block_type.outputs.at(0)!,
        },
        {
          block: textOutputBlockConfig,
          input: textOutputBlockConfig.block_type.inputs.at(0)!,
        },
      ),
    ).toEqual({
      blocks: [textInputBlockConfig, textOutputBlockConfig],
    });
  });
});

describe(PipelineGraph.getNodes, () => {
  test('returns all nodes', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [
        textInputBlockConfig,
        {
          ...textOutputBlockConfig,
          opts: { input: `${textInputBlockConfig.name}:output` },
        },
      ],
    };
    expect(PipelineGraph.getNodes(pipeline)).toMatchSnapshot();
  });
});

describe(PipelineGraph.getEdges, () => {
  test('returns all edges', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [
        textInputBlockConfig,
        {
          ...textOutputBlockConfig,
          opts: { input: `${textInputBlockConfig.name}:output` },
        },
      ],
    };
    expect(PipelineGraph.getEdges(pipeline)).toMatchSnapshot();
  });
});

describe(PipelineGraph.removeBlock, () => {
  test('removes existing block', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [textInputBlockConfig],
    };
    expect(PipelineGraph.removeBlock(pipeline, textInputBlockConfig)).toEqual({
      blocks: [],
    });
  });

  test('removes inputs connected to block', () => {
    const pipeline: PipelineGraph.IPipelineConfig = {
      blocks: [
        textInputBlockConfig,
        {
          ...textOutputBlockConfig,
          opts: { input: `${textInputBlockConfig.name}:output` },
        },
      ],
    };
    expect(PipelineGraph.removeBlock(pipeline, textInputBlockConfig)).toEqual({
      blocks: [textOutputBlockConfig],
    });
  });
});

const textInputBlockConfig: IBlockConfig = {
  opts: {},
  type: 'text_input',
  name: 'text_input',
  block_type: {
    inputs: [{ name: 'input', public: true, type: 'text' }],
    outputs: [{ name: 'output', public: false, type: 'text' }],
    schema: '',
    type: 'text_input',
  },
};

const textOutputBlockConfig: IBlockConfig = {
  opts: {},
  type: 'text_output',
  name: 'text_output',
  block_type: {
    inputs: [{ name: 'input', public: false, type: 'text' }],
    outputs: [{ name: 'output', public: true, type: 'text' }],
    schema: '',
    type: 'text_output',
  },
};

const audioInputBlockConfig: IBlockConfig = {
  opts: {},
  type: 'audio_input',
  name: 'audio_input',
  block_type: {
    inputs: [{ name: 'input', public: true, type: 'audio' }],
    outputs: [{ name: 'output', public: false, type: 'audio' }],
    schema: '',
    type: 'audio_input',
  },
};
