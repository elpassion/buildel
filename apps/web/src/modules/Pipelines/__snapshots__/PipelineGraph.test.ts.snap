// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getEdges > returns all edges 1`] = `
[
  {
    "id": "text_input:output-text_output:input",
    "source": "text_input",
    "sourceHandle": "output",
    "target": "text_output",
    "targetHandle": "input",
  },
]
`;

exports[`getNodes > returns all nodes 1`] = `
[
  {
    "data": {
      "block_type": {
        "inputs": [
          {
            "name": "input",
            "public": true,
            "type": "text",
          },
        ],
        "outputs": [
          {
            "name": "output",
            "public": false,
            "type": "text",
          },
        ],
        "schema": "",
        "type": "text_input",
      },
      "inputs": [],
      "name": "text_input",
      "opts": {},
      "position": {
        "x": 100,
        "y": 100,
      },
      "type": "text_input",
    },
    "id": "text_input",
    "position": {
      "x": 100,
      "y": 100,
    },
    "type": "text_input",
  },
  {
    "data": {
      "block_type": {
        "inputs": [
          {
            "name": "input",
            "public": false,
            "type": "text",
          },
        ],
        "outputs": [
          {
            "name": "output",
            "public": true,
            "type": "text",
          },
        ],
        "schema": "",
        "type": "text_output",
      },
      "inputs": [],
      "name": "text_output",
      "opts": {
        "input": "text_input:output",
      },
      "position": {
        "x": 100,
        "y": 100,
      },
      "type": "text_output",
    },
    "id": "text_output",
    "position": {
      "x": 100,
      "y": 100,
    },
    "type": "text_output",
  },
]
`;
